blockchain/pool.go

```diff
@@ -1,630 +1,372 @@
+			if block == nil {
+				// Since block was not yet received it is counted in numPending, decrement.
+				pool.numPending--
+				pool.peers[peerID].numPending--
+			}
+			pool.rescheduleRequest(peerID, h)
+		}
+		peer.height = height
 	}
-}
 
-// RemovePeer removes the peer with peerID from the pool. If there's no peer
-// with peerID, function is a no-op.
-func (pool *BlockPool) RemovePeer(peerID p2p.ID) {
-	pool.mtx.Lock()
-	defer pool.mtx.Unlock()
+	pool.updateMaxPeerHeight()
 
-	pool.removePeer(peerID)
+	return nil
 }
 
-func (pool *BlockPool) removePeer(peerID p2p.ID) {
-	for _, requester := range pool.requesters {
-		if requester.getPeerID() == peerID {
-			requester.redo(peerID)
+// Stops the peer timer and deletes the peer. Recomputes the max peer height.
+func (pool *blockPool) deletePeer(peerID p2p.ID) {
+	if p, ok := pool.peers[peerID]; ok {
+		if p.timeout != nil {
+			p.timeout.Stop()
 		}
-	}
+		delete(pool.peers, peerID)
 
-	peer, ok := pool.peers[peerID]
-	if ok {
-		if peer.timeout != nil {
-			peer.timeout.Stop()
+		if p.height == pool.maxPeerHeight {
+			pool.updateMaxPeerHeight()
 		}
+	}
+}
 
-		delete(pool.peers, peerID)
+// Removes any blocks and requests associated with the peer and deletes the peer.
+// Also triggers new requests if blocks have been removed.
+func (pool *blockPool) removePeer(peerID p2p.ID, err error) {
+	pool.logger.Debug("removing peer", "peerID", peerID)
 
-		// Find a new peer with the biggest height and update maxPeerHeight if the
-		// peer's height was the biggest.
-		if peer.height == pool.maxPeerHeight {
-			pool.updateMaxPeerHeight()
+	peer := pool.peers[peerID]
+	if peer == nil {
+		return
+	}
+	// Reschedule the requests for all blocks waiting for the peer, or received and not processed yet.
+	for h, block := range pool.peers[peerID].blocks {
+		if block == nil {
+			pool.numPending--
 		}
+		pool.rescheduleRequest(peerID, h)
 	}
+	pool.deletePeer(peerID)
+
 }
 
-// If no peers are left, maxPeerHeight is set to 0.
-func (pool *BlockPool) updateMaxPeerHeight() {
-	var max int64
+// Called every time FSM advances its height.
+func (pool *blockPool) removeShortPeers() {
 	for _, peer := range pool.peers {
-		if peer.height > max {
-			max = peer.height
+		if peer.height < pool.height {
+			pool.removePeer(peer.id, nil)
 		}
 	}
-	pool.maxPeerHeight = max
 }
 
-// Pick an available peer with at least the given minHeight.
-// If no peers are available, returns nil.
-func (pool *BlockPool) pickIncrAvailablePeer(minHeight int64) *bpPeer {
-	pool.mtx.Lock()
-	defer pool.mtx.Unlock()
-
+func (pool *blockPool) removeBadPeers() {
+	pool.removeShortPeers()
 	for _, peer := range pool.peers {
-		if peer.didTimeout {
-			pool.removePeer(peer.id)
-			continue
-		}
-		if peer.numPending >= maxPendingRequestsPerPeer {
-			continue
-		}
-		if peer.height < minHeight {
-			continue
+		if err := peer.isGood(); err != nil {
```

Comment:
It seems that result of `peer.isGood` depends on the instant at which it is called. Maybe it would be better isolating failure detection in a separate module that would inform other part of the code in case it detects the failure.